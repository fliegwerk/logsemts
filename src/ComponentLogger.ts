import Style from './types/Style';

/**
 * A logger for a specific component like _API Client_, _Message Provider_, or similar
 *
 * @internal
 *
 * @example ```ts
 * const = new ComponentLogger((type, style, ...args) => {
 *     const componentStyle = new Style(
 *         'red' // text color
 *         componentBackgroundColor // background color
 *     );
 *
 *     logFunction(type, style, componentName, componentStyle, ...args);
 * );
 * ```
 */
export default class ComponentLogger {
	/**
	 * @param semanticLogFunction A semantic log function that can already process the type and style of the semantic log type.
	 */
	constructor(
		private semanticLogFunction: (
			type: string,
			style: Style,
			...args: any[]
		) => void
	) {}

	/**
	 * Log some debug information
	 * @param args information that gets logged
	 */
	debug(...args: any[]) {
		this.semanticLogFunction('DEBUG', new Style('lightgray', 'black'), ...args);
	}

	/**
	 * Log some success information
	 * @param args information that gets logged
	 */
	success(...args: any[]) {
		this.semanticLogFunction('SUCCESS', new Style('white', 'green'), ...args);
	}

	/**
	 * Log an error
	 * @param args error information that gets logged
	 */
	error(...args: any[]) {
		this.semanticLogFunction('ERROR', new Style('white', 'red'), ...args);
	}

	/**
	 * Log a warning
	 * @param args warning that gets logged
	 */
	warn(...args: any[]) {
		this.semanticLogFunction('WARN', new Style('white', 'orange'), ...args);
	}

	/**
	 * Log some information
	 * @param args information that gets logged
	 */
	info(...args: any[]) {
		this.semanticLogFunction('INFO', new Style('white', 'black'), ...args);
	}
}
