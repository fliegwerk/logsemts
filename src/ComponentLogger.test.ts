import { ComponentLogger } from './ComponentLogger';
import MockedFunction = jest.MockedFunction;

describe('ComponentLogger', () => {
	let logFn: MockedFunction<any>;
	let componentLogger: ComponentLogger;

	beforeEach(() => {
		logFn = jest.fn();
		componentLogger = new ComponentLogger(logFn);
	});

	(['debug', 'success', 'warn', 'error', 'info'] as const).forEach(
		functionName =>
			it(`should log ${functionName} messages`, () => {
				componentLogger[functionName]('Test');

				expect(logFn).toHaveBeenCalled();
				expect(logFn).toMatchSnapshot();
			})
	);
});
