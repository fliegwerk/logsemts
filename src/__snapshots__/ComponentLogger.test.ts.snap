// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentLogger should log debug messages 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "DEBUG",
      Object {
        "background": "black",
        "color": "lightgray",
        "css": "color: lightgray; background-color: black; font-weight: bold;",
      },
      "Test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ComponentLogger should log error messages 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "ERROR",
      Object {
        "background": "red",
        "color": "white",
        "css": "color: white; background-color: red; font-weight: bold;",
      },
      "Test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ComponentLogger should log info messages 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "INFO",
      Object {
        "background": "black",
        "color": "white",
        "css": "color: white; background-color: black; font-weight: bold;",
      },
      "Test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ComponentLogger should log success messages 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "SUCCESS",
      Object {
        "background": "green",
        "color": "white",
        "css": "color: white; background-color: green; font-weight: bold;",
      },
      "Test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ComponentLogger should log warn messages 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "WARN",
      Object {
        "background": "orange",
        "color": "white",
        "css": "color: white; background-color: orange; font-weight: bold;",
      },
      "Test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
